### ISSUE TITLE
Select @rid from CLUSTER:default memory leak

### ISSUE THREAD

**Issue by stoica-florian** [orientechnologies/orientdb#9186]
## Expected behavior  
When doing a select for a specific field liked of recordId, only the id to be fetched and processed, not the whole record.

## Actual behavior  
For some reason, it seems that  when selecting all the record ids also the data is loaded and kept in memory.

## Steps to reproduce  
Used gremlin java API(but this is not an API bug using orient db studio this happened as well)

`  OResultSet result = dbSession.command("select @rid from CLUSTER:default", parameters);
        result.forEachRemaining(results::add);
        result.close();`
this will fill up the memory and  I was unable to release it without server restart. Doing a delete  like 
`delete from CLUSTER:default where @rid not in [#3:0]` will have the same behaviour.

Tried using javascript command, closing dbSession closing factory...you name it....no positive result.

In short, I want to delete unused ORecordBytes from the default cluster (truncate does not allow **not** [...ids])

**Comment by luigidellaquila** [comment]
Hi @stoica-florian 

There are two different memory management models in OrientDB, one is off-heap for the disl cache, the other one is in heap for all the other operations (query execution, transactions and so on).

The disk cache is just a mirror of the disk structures in RAM, to speed up read operations and avoid disk access in the average case. Disk cache is not supposed to be emptied, so when you load a big cluster, you'll typically end up filling it. The eviction strategy takes into consideration the page access (ie. LRU) and takes place when there is no more room in the cache itself.

The heap is used in different ways depending on the case, eg. 
- the transactions are kept in heap until they are committed, then the memory is released; a big transaction can consume a big amout of heap, but after commit it is released
- queries are also executed in heap, but they are streamed, ie. a big result set is split in multiple small 
 segments that are fetched one by one from disk/cache, returned to the client and then released, so there is typically no chance to fill the heap with a normal query (an exception to this are aggregate operations (eg. GROUP BY) and sometimes ORDER BY, in particular when they are not optimized using indexes)

I suppose in your case the memory that is not being released is the off-heap disk cache, that's an absolutely normal behaviour. You can tune the maximum disk-cache size anyway, see  https://orientdb.com/docs/3.0.x/tuning/Performance-Tuning.html#server-and-embedded-settings

I hope it helps

Thanks

Luigi

**Comment by stoica-florian** [comment]
Hi @luigidellaquila ,
Thanks for the quick response.
I've changed the disk-cache size with java option, still the same behavior.
Wrapped around a transaction committed the transaction still the same behavior.

Can you please tell me why is it normal to load the whole record when in fact only the id is needed, or why is it kept if only that information is retrieved.

Can you provided some way in which this will work? In the current case, I'm deleting around 100K records that are not used it takes around 10GB of RAM.

Can the id's be retrieved in a different fashion?

Thanks,
Florian

**Comment by luigidellaquila** [comment]
Hi Florian

Are you doing it in a single transaction?  Can you post some sample code?

Thanks

Luigi

**Comment by stoica-florian** [comment]
Here is the latest try:

```
 dbSession.begin(OTransaction.TXTYPE.OPTIMISTIC);
        OTransaction transaction = dbSession.getTransaction();
        OResultSet result = transaction.getDatabase().query("select @rid.asString() from CLUSTER:default", new HashMap<>());
        while (result.hasNext()) {
            OResultInternal ri = (OResultInternal) result.next();
            results.add(ri.getProperty("@rid.asString()"));
        }
        result.close();
        transaction.commit(true);
        transaction.close();
        dbSession.commit(true);
```

**Comment by luigidellaquila** [comment]
Hi @stoica-florian 

In this case, the problem is that the transaction is all kept in memory before being committed.

I'd suggest to change the code as follows and batch the commits:

```
        dbSession.begin(OTransaction.TXTYPE.OPTIMISTIC);
        OResultSet result = dbSession.query("select @rid.asString() from CLUSTER:default", new HashMap<>());
        int i = 0;
        while (result.hasNext()) {
            OResultInternal ri = (OResultInternal) result.next();
            results.add(ri.getProperty("@rid.asString()"));
            if(i % 500 == 0) {
                dbSession.commit();
                dbSession.begin();
            }
            i++;
        }
        result.close();
        dbSession.commit(true);
```

(btw, you don't need to get the TX explicitly, you can do all the job using the ODatabaseSession only)

Thanks

Luigi

**Comment by luigidellaquila** [comment]
oh wait... I was having a closer look at the code, you are not deleting the elements here... why do you need a tx at all?

Thanks

Luigi

**Comment by stoica-florian** [comment]
this is the part that occupies the too much RAM memory at this point I'm just trying anything initially it was without transaction directly with obsession. I also tried the above approach and still the memory is not freed at all.

Thanks,
Florian

**Comment by stoica-florian** [comment]
Even if I close the db session the memory is not freed.

**Comment by luigidellaquila** [comment]
Hi @stoica-florian 

When you close the db, all the heap memory is freed for sure.
Please consider that you are also accumulating RIDs in a collectin:

```
 results.add(ri.getProperty("@rid.asString()"));
```

That memory won't be freed of course

Thanks

Luigi

**Comment by stoica-florian** [comment]
Hi @luigidellaquila 

```
 String embeddedURL = "embedded:" + Directories.getDBFolder().resolve(dbName).toString();
        BaseConfiguration newConfig = new BaseConfiguration();
        newConfig.setProperty(CONFIG_URL, embeddedURL);
        newConfig.setProperty(CONFIG_USER, "admin");
        newConfig.setProperty(CONFIG_PASS, "admin");
        newConfig.setProperty(CONFIG_POOL_SIZE, 50);

        OrientGraphFactory factory = new OrientGraphFactory(newConfig);
        ODatabaseDocument databaseDocument = factory.getDatabase(true, false);
        OResultSet result = databaseDocument.query("select @rid.asString() from CLUSTER:default", new HashMap<>());
        int i = 0;
        while (result.hasNext()) {
            OResultInternal ri = (OResultInternal) result.next();
            results.add(ri.getProperty("@rid.asString()"));
            if(i % 500 == 0) {
                databaseDocument.commit();
                databaseDocument.begin();
            }
            i++;
        }
        result.close();
        databaseDocument.commit(true);
        databaseDocument.close();
        factory.close();
```
This code does not free the memory.
RIDs are just strings that won't keep 10GB occupied.

Thanks,
Florian

**Comment by luigidellaquila** [comment]
Hi @stoica-florian 

Can you please post the jvm parameters you pass at startup?

IMHO you should be able to run this with much much less than 10GB...

Thanks

Luigi

**Comment by stoica-florian** [comment]
`java -agentlib:jdwp=transport=dt_socket,address=127.0.0.1:42489,suspend=y,server=n -Dstorage.diskCache.bufferSize=7200 -XX:TieredStopAtLevel=1 -noverify -Dspring.output.ansi.enabled=always -Dcom.sun.management.jmxremote -Dspring.jmx.enabled=true -Dspring.liveBeansView.mbeanDomain -Dspring.application.admin.enabled=true -javaagent:/snap/intellij-idea-ultimate/212/plugins/java/lib/rt/debugger-agent.jar -Dfile.encoding=UTF-8 -classpath `

Maybe I was misunderstood...the issue is that when iterating over the result set with .next RAM is filled incrementally until around 10GB (if I had more data it would go up more) and not even after closing all things the memoir is not freed until the java server is stopped.

Thanks,
Florian

**Comment by luigidellaquila** [comment]
Hi @stoica-florian 

I see you have 7GB of disk cache, that is not supposed to be released (it will be there for future reads, without the need to access disk again).
The rest is heap memory, that is under the control of the JVM and for sure is not released immediately.
I'd say in general it's an expected behaviour. 

The only thing I can suggest, if memory footprint is a concern for you, is to reduce the disk cache and give an explicit -Xmx to limit the heap.

Thanks

Luigi

**Comment by stoica-florian** [comment]
Thanks for the long talk, the combination of the two worked. (I've added the 7GB because of the first reply of this issue initially no argument was there regarding disk-cache).

I'm still inclined that there is an issue there because for selecting Ids the whole record is loaded, but as long as this workaround works, I'm gonna close the Issue and thanks again for the fast response and great support.
